### PR TITLE
feat: Sending configuration to mixpanel

### LINKED ISSUE #1503: [Mixpanel] Collect configuration statistics
**Author the user story for this feature**

As a flank developer, I want to know what options the community using so I can understand how the community uses Flank.

**Describe the solution**

1. User should be registered by project id.
2. Flank should check what properties are set and send it as a single event. 
3. Properties like ```app``` or ```environment variables``` should be anonymized.
4. Flank shouldn't send events when running by integration and unit tests.
5. Maybe we should add a flag to turn off collecting data?

### PR DESCRIPTION
**Pull request by adamfilipow92**
Fixes #1503 

1. Flank sending configuration to mixpanel
2. If integration tests or unit sending configuration disabled
3. Flag to disable sending statistics works
4. ```environment-variables```, paths to files anonymized (maybe something more?)

## Checklist

- [X] Sending usage statistics to mixpanel
- [x] Flag to disable usage statistic
- [X] Disable usage statistic in tests
- [X] Anonymized sensitive data
- [X] Unit tested
- [X] Integration tests updated
- [X] respects the same analytics opt out as gcloud CLI.

### CODE REVIEW

**Comment by pawelpasterz** [comment]
I think we should not send statistics from developer environment, only from production one (our, local, manual tests could spoil statistics)

**Comment by pawelpasterz** [comment]
> I think we should not send statistics from developer environment (our, local, manual tests could spoil statistics). DIsabling for IT or UT is great but I think it won't be enough. Statistics should be only from official releases?
> This is an open question.

On the second thought, it should be difficult to filter `local_snapshot` statistics so it should not be an issue (at least I think)

**Comment by piotradamczyk5** [comment]
> > I think we should not send statistics from developer environment (our, local, manual tests could spoil statistics). DIsabling for IT or UT is great but I think it won't be enough. Statistics should be only from official releases?
> > This is an open question.
> 
> On the second thought, it should be possible to filter `local_snapshot` statistics so it should not be an issue (at least I think)

we could make a flag and store it in FlankProperties 😃

**Comment by bootstraponline** [comment]
I think the uuid opt out should also apply to mixpanel

```
Flank respects the same analytics opt out as gcloud CLI.

echo "DISABLED" > ~/.gsutil/analytics-uuid
```
- https://github.com/Flank/flank/blob/master/docs/error_monitoring.md#disable-bugsnag

File: test_runner/src/main/kotlin/ftl/analytics/SendUsageStatistics.kt [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import ftl.args.IArgs
+import ftl.args.IosArgs
+import ftl.args.blockSendingUsageStatistics
+import ftl.util.isGoogleAnalyticsDisabled
+
+fun AndroidArgs.sendConfiguration() = sendConfiguration(events = createEventMap())
+
+fun IosArgs.sendConfiguration() = sendConfiguration(events = createEventMap())
+
+fun IArgs.sendConfiguration(events: Map<String, Any?>, rootPath: String = System.getProperty("user.home")) =
```
Comment (piotradamczyk5) [inline test_runner/src/main/kotlin/ftl/analytics/SendUsageStatistics.kt:13]:
```suggestion
fun IArgs.sendConfiguration(events: Map<String, Any?>, rootPath: String = userHome) =
```

File: test_runner/src/main/kotlin/ftl/analytics/StatisticDataFilters.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import kotlin.reflect.KClass
+
+private const val ANONYMIZE_VALUE = "..."
+
+annotation class IgnoreInStatistics
+annotation class AnonymizeInStatistics
+
+private val classesForStatistics = listOf(IArgs::class, AndroidArgs::class, IosArgs::class)
+
+private fun KClass<*>.ignoredMembersForStatistics() = findMembersWithAnnotation(IgnoreInStatistics::class)
```
Comment (piotradamczyk5) [inline test_runner/src/main/kotlin/ftl/analytics/StatisticDataFilters.kt:15]: please move below `keysToRemove`

File: test_runner/src/main/kotlin/ftl/analytics/StatisticDataFilters.kt [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+private fun KClass<*>.anonymousMembersForStatistics() = findMembersWithAnnotation(AnonymizeInStatistics::class)
```
Comment (piotradamczyk5) [inline test_runner/src/main/kotlin/ftl/analytics/StatisticDataFilters.kt:16]: please move below `keysToAnonymize `

File: test_runner/src/main/kotlin/ftl/analytics/UsageStatisticsClient.kt [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        addModule(kotlinModule())
+    }
+}
+
+@VisibleForTesting
+internal fun JSONObject.send() = apiClient.sendMessage(this)
+
+internal fun Map<String, Any?>.toEvent(projectId: String) =
+    (this + Pair(SESSION_ID, sessionId)).run {
+        messageBuilder.event(projectId, CONFIGURATION_KEY, JSONObject(this))
```
Comment (piotradamczyk5) [inline test_runner/src/main/kotlin/ftl/analytics/UsageStatisticsClient.kt:34]:
```suggestion
        messageBuilder.event(projectId, CONFIGURATION_KEY, toJSONObject())
```

**Comment by piotradamczyk5** [comment]
Please remeber to update `flank.yml`, `flank.ios.yml`, `docs/index.md` etc with new option

File: test_runner/src/main/kotlin/ftl/analytics/StatisticDataFilters.kt (hunk 2 above)
Comment (adamfilipow92) [reply test_runner/src/main/kotlin/ftl/analytics/StatisticDataFilters.kt:15]: Changed thanks! 👍

File: test_runner/src/main/kotlin/ftl/analytics/StatisticDataFilters.kt (hunk 3 above)
Comment (adamfilipow92) [reply test_runner/src/main/kotlin/ftl/analytics/StatisticDataFilters.kt:16]: Changed thanks! 👍

File: test_runner/src/main/kotlin/ftl/analytics/StatisticDataFilters.kt [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+private val classesForStatistics = listOf(IArgs::class, AndroidArgs::class, IosArgs::class)
+
+private const val ANONYMIZE_VALUE = "..."
+
+private fun KClass<*>.findMembersWithAnnotation(annotationType: KClass<*>) = members.filter { member ->
+    member.annotations.any { annotation -> annotation.annotationClass == annotationType }
+}.map {
+    it.name
+}
```
Comment (jan-gogo) [inline test_runner/src/main/kotlin/ftl/analytics/StatisticDataFilters.kt:31]:
```suggestion
private fun findMembersWithAnnotation(
    annotationType: KClass<*>
): KClass<*>.() -> List<String> = {
    members.filter { member ->
        member.annotations.any { annotation -> annotation.annotationClass == annotationType }
    }.map {
        it.name
    }
}
```

**Review by jan-gogo — COMMENTED** [review]
The implementation looks well! 👍 I have added a few suggestions but all related to one thing.

File: test_runner/src/main/kotlin/ftl/analytics/StatisticDataFilters.kt [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+annotation class AnonymizeInStatistics
+
+internal val keysToRemove by lazy {
+    classesForStatistics.map { it.ignoredMembersForStatistics() }.flatten()
+}
+
+private fun KClass<*>.ignoredMembersForStatistics() = findMembersWithAnnotation(IgnoreInStatistics::class)
+
+internal val keysToAnonymize by lazy {
+    classesForStatistics.map { it.anonymousMembersForStatistics() }.flatten()
```
[the remark this answers is not in the document]
Comment (piotradamczyk5) [reply test_runner/src/main/kotlin/ftl/analytics/StatisticDataFilters.kt:18]: applied

**Review by piotradamczyk5 — APPROVED** [review]

**Review by jan-gogo — APPROVED** [review]

**Pull request merged**